scope.go

```diff
@@ -182,52 +202,164 @@ func newRootScope(opts ScopeOptions, interval time.Duration) *scope {
 }
 
 // report dumps all aggregated stats into the reporter. Should be called automatically by the root scope periodically.
-func (s *scope) report(r StatsReporter) {
-	s.cm.RLock()
-	for name, counter := range s.counters {
-		counter.report(s.fullyQualifiedName(name), s.tags, r)
+func (s *scope) report(r StatsReporter) bool {
+	var numMetrics, i int
+
+	s.cm.Lock()
```

Comment:
is `Lock` vs `RLock` going to have any negative impact?

should we instead under the rlock, come up with a list of counters to delete, and then do a separate `Lock` to just delete those items?

Reply:
Yeah I thought about doing that - the main negative impact is going to be during a report call and users doing things like scope.Counter("myCounter").Inc(1) since those will be blocked, but those calls are already taking a complicated key generation and map lookup cost and are in the microseconds already. For users who care about perf, they're already holding a ref to the counter itself, so won't be impacted by this lock.

The problem with doing RLock first and deleting counters later is that it introduces a race condition such that you can have a metric marked as expired from the report call in line 211 and inserted into the to be deleted slice, and then marked as non expired if it gets updated before the Lock is acquired. This could result in a situation where the increment is lost.

I guess we can do a final check for expiration of the counter before doing the delete in the lock and if it's no longer expired, skip it. Will change.